frame/vesting/src/lib.rs

```diff
@@ -61,41 +60,139 @@ use frame_support::traits::{
+mod vesting_info {
+	use super::*;
+	/// Struct to encode the vesting schedule of an individual account.
+	#[derive(Encode, Decode, Copy, Clone, PartialEq, Eq, RuntimeDebug)]
+	pub struct VestingInfo<Balance, BlockNumber> {
+		/// Locked amount at genesis.
+		locked: Balance,
+		/// Amount that gets unlocked every block after `starting_block`.
+		per_block: Balance,
+		/// Starting block for unlocking(vesting).
+		starting_block: BlockNumber,
+	}
+
+	impl<Balance: AtLeast32BitUnsigned + Copy, BlockNumber: AtLeast32BitUnsigned + Copy>
+		VestingInfo<Balance, BlockNumber>
+	{
+		/// Instantiate a new `VestingInfo` and validate parameters
+		pub fn try_new<T: Config>(
+			locked: Balance,
+			per_block: Balance,
+			starting_block: BlockNumber,
+		) -> Result<VestingInfo<Balance, BlockNumber>, Error<T>> {
+			Self::validate_params(locked, per_block, starting_block)?;
+			let per_block = if per_block > locked { locked } else { per_block };
+			Ok(VestingInfo { locked, per_block, starting_block })
+		}
+
+		/// Validate parameters for `VestingInfo`.
+		pub fn validate_params<T: Config>(
+			locked: Balance,
+			per_block: Balance,
+			_starting_block: BlockNumber,
+		) -> Result<(), Error<T>> {
+			ensure!(!locked.is_zero() && !per_block.is_zero(), Error::<T>::InvalidScheduleParams);
+			let min_transfer: u32 = T::MinVestedTransfer::get().try_into().unwrap_or(u32::MAX);
+			let min_transfer = Balance::from(min_transfer);
+			ensure!(locked >= min_transfer, Error::<T>::AmountLow);
+			Ok(())
+		}
+
+		/// Instantiate a new `VestingInfo` without param validation. Useful for
+		/// mocking bad inputs in testing.
+		pub fn unsafe_new(
+			locked: Balance,
+			per_block: Balance,
+			starting_block: BlockNumber,
+		) -> VestingInfo<Balance, BlockNumber> {
+			VestingInfo { locked, per_block, starting_block }
+		}
+
+		/// Locked amount at genesis.
+		pub fn locked(&self) -> Balance {
+			self.locked
+		}
+
+		/// Amount that gets unlocked every block after `starting_block`.
+		pub fn per_block(&self) -> Balance {
+			self.per_block
+		}
+
+		/// Starting block for unlocking(vesting).
+		pub fn starting_block(&self) -> BlockNumber {
+			self.starting_block
+		}
+
+		/// Amount locked at block `n`.
+		pub fn locked_at<BlockNumberToBalance: Convert<BlockNumber, Balance>>(
+			&self,
+			n: BlockNumber,
+		) -> Balance {
+			// Number of blocks that count toward vesting
+			// Saturating to 0 when n < starting_block
+			let vested_block_count = n.saturating_sub(self.starting_block);
+			let vested_block_count = BlockNumberToBalance::convert(vested_block_count);
+			// Return amount that is still locked in vesting
+			let maybe_balance = vested_block_count.checked_mul(&self.per_block);
+			if let Some(balance) = maybe_balance {
+				self.locked.saturating_sub(balance)
+			} else {
+				Zero::zero()
+			}
+		}
+
+		/// Block number at which the schedule ends
+		pub fn ending_block<BlockNumberToBalance: Convert<BlockNumber, Balance>>(&self) -> Balance {
```

Comment:
This signature is strange: the function name and documentation suggest that it'll return a `BlockNumber`, but it returns a `Balance` instead. Might be worth converting the result back.

Reply:
agreed.

Reply:
Yup I agree. This also brings up the edge case where `locked / duration > BlockNumber::MAX`, which I think we should validate new schedules for and check on here

Reply:
I am going to leave this for now and come back to it once I refactor some of the core logic for filtering out vesting schedules and updating locks

Reply:
I am thinking it is best to leave the behaviour unchanged and just change the fn name to reflect it.
Reasoning:
- We only use this method once, and we need the return value as a balance.
- Considering there is a one way generic conversion fn as a type param, we can't be sure of the intended behavior when converting the other direction. 

We could also consider reducing this fn's visibility to `pub(crate)`.

Reply:
I changed the fn's name to `ending_block_as_blocknumber` e67c6c5